<div style="text-align:center">
<img src="img/intro_correspondre.png">
</div>


*** CREATION IN PROGRESS ***<br>
Notebook created by GITHUB ESTELLE DERRIEN

## Summary

15. <b> Modeling of a multiperiod production and storage problem </b>
    - 15.1 Taïwan university instance 
    - 15.2 Mathematical modeling
    - 15.3 Python Pulp solution and comparison
    - 15.4 Columbia university instance
    - 15.5 Mathematical modeling
    - 15.6 Python Pulp solution and comparison
16. <b> Modeling of a multiperiod cash flow matching </b>
    - 15.1 Columbia university instance 
    - 15.2 Mathematical modeling
    - 15.3 Python Pulp solution and comparison
    - 15.4 Washington university instance
    - 15.5 Mathematical modeling
    - 15.6 Python Pulp solution and comparison
17. <b>Modeling of an optimal employees time shift scheduling problem</b>
    - 16.1 Instance - Microsoft problem : "Using Solver to schedule your workforce".
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution
18. <b>Modeling of a linear portfolio optimization problem</b>
    - 16.1 Instance by Excel easy
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution
    - 16.1 Instance by Mastering python for finance p.96
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution
    - 16.1 Instance by Kenneth A Barker
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution
    - 16.1 Instance by Andrew Hershy
    - 16.2 Mathematical modeling
    - 16.3 Python Pulp solution
19. <b>Modeling of a facility location problem</b>
    - 17.1 Instance - Taïwan university problem
    - 17.2 Mathematical modeling
    - 17.3 Python Pulp solution and Chang Yaochen code analysis
20. <b> Modeling of a single machine scheduling problem</b>
    - 18.1 Instance
    - 18.2 Mathematical modeling
    - 18.3 Python Pulp solution
    - 18.4 Single machine, parallels ,flow shop and job shop differences.
21. <b> Modeling a chemistry combinatorial optimization  </b>
    - Presentation of the Mac Graw Book problem BOILER/TURBO-GENERATOR SYSTEM
    - Mathematical modeling
    - Python Pulp or Cplex solution
22. <b> Modeling a petroleum combinatorial optimization </b>
    - Presentation and Kirchhoff constraints
    - Mathematical modeling
    - Python Pulp or Cplex solution
23. <b> Modeling an uncapacited transport problem </b>
    - Presentation 
    - Mathematical modeling
    - Python Pulp or Cplex solution
24. <b> Modeling a capacited transport problem  </b>
    - Presentation of the Daskin problem (2013)
    - Mathematical modeling
    - Python Pulp or Cplex solution
        - Using hard values
        - Using iterations
25. <b> Modeling a bin packing problem  </b>
    - Presentation and differences between 2d and 3d
    - Mathematical modeling
    - Python Pulp or Cplex solution
26. <b> Modeling a traveling salesman problem  </b>
    - Presentation 
    - Mathematical modeling
    - Python Pulp or Cplex solution
27. <b> Modeling a shortest path problem  </b>
    - Presentation of https://blog.sommer-forst.de/2013/04/06/solving-the-minimum-cost-flow-problem-4-pulp/
    - https://transport-systems.imperial.ac.uk/tf/60008_21/n3_6_shortest_paths_pulp.html
    - Mathematical modeling
    - Python Pulp or Cplex solution
28. <b> Modeling of a multiple objectives problem</b>
    - Presentation of https://www.supplychaindataanalytics.com/multi-objective-linear-optimization-with-pulp-in-python/
    - Mathematical modeling
    - Python Pulp or Cplex solution
29. <b> Getting familiar with sensitivity analysis</b>
    - Displaying Reduced costs
    - Displaying The Shadow Price
    - Displaying Slacks variables
30. <b> Getting familiar with variables relaxations</b>

31. <b> Linearizations </b>
    - Linearization technics using the AIMMS modeling guide and Ling-Chieh Kung technics
    - Objective linearizations
    - Constraints linearizations
    - Examples using Python solvers


32. <b> Apendix </b> 
- Importing a data instance from an Excel file
- Importing a data instance from a Database

# <b>15. Modeling of a multiperiod production and storage problem </b>

<b>15.1 Instance</b>

The manager predicts that we'll sell 100, 150, 200, and 170 on days 1, 2, 3, and 4 of the next week.

Production costs vary depending on the day of the week:
9th, 12th, 10th, and 12th on days 1, 2, 3, and 4 of the next week.

The storage cost is €1 per item stored, so if we keep it for 4 days, that's a cost of €4.

What is the best production plan to minimize costs and not have too much inventory?

<b>15.2 Math model</b><br>
Step 1 : Simplified/hard coded model :
<img src="img/01. Formulation du L.P.png"></img>


<b>15.3 Python Pulp solution</b>
<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

Hard coded version :

In [3]:
# -----------------------------------
# Import Pulp
# -----------------------------------
import pulp

# -----------------------------------
# Type of problem
# -----------------------------------
problem = pulp.LpProblem("Problem",pulp.LpMinimize)


# -----------------------------------
# Decision variables
# -----------------------------------
# Items production per day 1,2,3,4
P1 = pulp.LpVariable('Prodution1', lowBound=0, cat='Integer')
P2 = pulp.LpVariable('Prodution2', lowBound=0, cat='Integer')
P3 = pulp.LpVariable('Prodution3', lowBound=0, cat='Integer')
P4 = pulp.LpVariable('Prodution4', lowBound=0, cat='Integer')


# Inventory per day 1,2,3,4
S1 = pulp.LpVariable('Stock1', lowBound=0, cat='Integer')
S2 = pulp.LpVariable('Stock2', lowBound=0, cat='Integer')
S3 = pulp.LpVariable('Stock3', lowBound=0, cat='Integer')
S4 = pulp.LpVariable('Stock4', lowBound=0, cat='Integer')

# -----------------------------------
# Math model
# -----------------------------------

# Objective function : minimizing prod costs + inventory costs
problem += (9*P1 + 12*P2 + 10*P3 + 12*P4) + (S1 + S2 + S3 + S4) 

# Inventory balancing constraints
problem += P1 - 100 == S1
problem += S1 + P2 - 150 == S2
problem += S2 + P3 - 200 == S3
problem += S3 + P4 - 170 == S4

problem += P1 >= 0
problem += P2 >= 0 
problem += P3 >= 0 
problem += P4 >= 0 

# -----------------------------------
# Solution
# -----------------------------------
result = problem.solve()

# Printing solution
print("---First day---")
print("Production:" ,pulp.value(P1))
print("Inventory:" ,pulp.value(S1))


print("---2nd day---")
print("Production:" ,pulp.value(P2))
print("Inventory:" ,pulp.value(S2))

print("---3rd day---")
print("Production:" ,pulp.value(P3))
print("Inventory:" ,pulp.value(S3))

print("---4th day---")
print("Production:" ,pulp.value(P4))
print("Inventory:" ,pulp.value(S4))

print("Total cost:" ,pulp.value(problem.objective))



---First day---
Production: 250.0
Inventory: 150.0
---2nd day---
Production: 0.0
Inventory: 0.0
---3rd day---
Production: 370.0
Inventory: 170.0
---4th day---
Production: 0.0
Inventory: 0.0
Total cost: 6270.0


Our multi-period interval here is 4 days.

The solver calculates what needs to be produced and put into inventory in order to end up with no inventory on hand at the end of these 4 days, while still meeting demand.

On the first day, we produce 250 units, which meets the manager's forecast of selling 100 units on the first day, and we keep 150 units in inventory for the second day.

On the third day, we produce 370 units, still with the aim of meeting the manager's forecast for the following 2 days, and we keep 170 units in inventory.

On the fourth day, we have minimized our inventory and our carrying costs, and we have nothing left.

<b>15.4 Python Pulp solution ( Using instance and iterations)</b>
<div style="text-align:center">
<img src="img/small-pulp.png" width="100">
</div>

In [4]:
# -----------------------------------
# Import Pulp
# -----------------------------------
from pulp import *

# -----------------------------------
# Type of problem
# -----------------------------------
model = LpProblem("Minimiser_le_cout_production",LpMinimize)

# -----------------------------------
# Instance
# -----------------------------------
# Define the production and storage costs for days 1, 2, 3, and 4; in quarters, it's the number of days, i.e., 4.
# Define production cost, inventory cost, and demand.
# Define the production costs per period, the inventory cost, and the demand reported by the sales department.
quaters = list(range(4))
cout_production=[9,12,10,12]
cout_stockage=[1,1,1,1]
demande=[100,150,200,170]

# -----------------------------------
# Decision variables
# -----------------------------------
x = LpVariable.dicts('quater_prod_', quaters,lowBound=0, cat='Continuous')
y = LpVariable.dicts('quater_stock_', quaters,lowBound=0, cat='Continuous')

# -----------------------------------
# Math model
# -----------------------------------
# Objective
model += lpSum([cout_production[i]*x[i] for i in quaters]) + lpSum([cout_stockage[i]*y[i] for i in quaters])

# Constraints

# Production-capacity constraints
for i in quaters:
    model.addConstraint(x[i]<=3000)

# Inventory-balance constraints
model.addConstraint(x[0] - y[0] == demande[0]) # (day 1)

for i in quaters[1:]:
    model.addConstraint(x[i] - y[i] + y[i-1] == demande[i]) # (days 2, 3, 4) 

# -----------------------------------
# Solution
# -----------------------------------
model.solve()

for v in model.variables():
    print(v.name, "=", v.varValue)
    
print("Total minimized cost =  ", value(model.objective))

quater_prod__0 = 250.0
quater_prod__1 = 0.0
quater_prod__2 = 370.0
quater_prod__3 = 0.0
quater_stock__0 = 150.0
quater_stock__1 = 0.0
quater_stock__2 = 170.0
quater_stock__3 = 0.0
Total minimized cost =   6270.0
